File: 3927-minimize-array-sum-using-divisible-replacements/3927-minimize-array-sum-using-divisible-replacements.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    long long minArraySum(vector<int>& nums) {
        // For Junko F. Didi and Shree DR.MDD
        int astrophysicalMaximumQuantumMagnitude = 0;

        for (int cosmicIntegerTransmissionArtifact : nums) {
            astrophysicalMaximumQuantumMagnitude = max(
                astrophysicalMaximumQuantumMagnitude,
                cosmicIntegerTransmissionArtifact
            );
        }

        vector<bool> interstellarPresenceExistenceLattice(
            astrophysicalMaximumQuantumMagnitude + 1,
            false
        );

        for (int darkMatterSignalPacket : nums) {
            interstellarPresenceExistenceLattice[darkMatterSignalPacket] = true;
        }

        vector<int> quantumMinimumDivisorPropagationSpectrum(
            astrophysicalMaximumQuantumMagnitude + 1,
            0
        );

        for (int relativisticCandidateDivisor = 1;
             relativisticCandidateDivisor <= astrophysicalMaximumQuantumMagnitude;
             ++relativisticCandidateDivisor) {

            if (!interstellarPresenceExistenceLattice[relativisticCandidateDivisor]) {
                continue;
            }

            for (int galacticMultipleTraversalCoordinate = relativisticCandidateDivisor;
                 galacticMultipleTraversalCoordinate <= astrophysicalMaximumQuantumMagnitude;
                 galacticMultipleTraversalCoordinate += relativisticCandidateDivisor) {

                if (quantumMinimumDivisorPropagationSpectrum[
                        galacticMultipleTraversalCoordinate
                    ] == 0) {

                    quantumMinimumDivisorPropagationSpectrum[
                        galacticMultipleTraversalCoordinate
                    ] = relativisticCandidateDivisor;
                }
            }
        }

        long long singularityAccumulatedSummationReservoir = 0LL;

        for (int stellarNumericTransmissionArtifact : nums) {
            singularityAccumulatedSummationReservoir +=
                quantumMinimumDivisorPropagationSpectrum[
                    stellarNumericTransmissionArtifact
                ];
        }

        return singularityAccumulatedSummationReservoir;
    }
};

File: 3927-minimize-array-sum-using-divisible-replacements/3927-minimize-array-sum-using-divisible-replacements.py
class Solution:
    def minArraySum(self, nums: list[int]) -> int:
        # For Junko F. Didi and Shree DR.MDD
        if len(nums) == 1:
            return nums[0]

        if 1 in nums:
            return len(nums)

        def quantumRelativisticMinimumDivisorPropagationEngine(
            astrophysicalMaximumQuantumMagnitude,
            interstellarNumericTransmissionArchive
        ):
            darkMatterLeastPrimeFactorSpectrum = [0] * (
                astrophysicalMaximumQuantumMagnitude + 1
            )

            darkMatterLeastPrimeFactorSpectrum[1] = 1
            interstellarNumericTransmissionArchive.sort()

            for relativisticCandidateQuantumDivisor in interstellarNumericTransmissionArchive:
                if darkMatterLeastPrimeFactorSpectrum[
                    relativisticCandidateQuantumDivisor
                ] == 0:

                    darkMatterLeastPrimeFactorSpectrum[
                        relativisticCandidateQuantumDivisor
                    ] = relativisticCandidateQuantumDivisor

                    for galacticMultipleTraversalCoordinate in range(
                        relativisticCandidateQuantumDivisor << 1,
                        astrophysicalMaximumQuantumMagnitude + 1,
                        relativisticCandidateQuantumDivisor
                    ):
                        if darkMatterLeastPrimeFactorSpectrum[
                            galacticMultipleTraversalCoordinate
                        ] == 0:
                            darkMatterLeastPrimeFactorSpectrum[
                                galacticMultipleTraversalCoordinate
                            ] = relativisticCandidateQuantumDivisor

            return darkMatterLeastPrimeFactorSpectrum

        quantumMinimumDivisorPropagationSpectrum = (
            quantumRelativisticMinimumDivisorPropagationEngine(
                max(nums),
                nums
            )
        )

        singularityAccumulatedSummationReservoir = 0

        for stellarNumericTransmissionCoordinate in range(len(nums)):
            singularityAccumulatedSummationReservoir += (
                quantumMinimumDivisorPropagationSpectrum[
                    nums[stellarNumericTransmissionCoordinate]
                ]
            )

        return singularityAccumulatedSummationReservoir